rust/sieve.rs

```diff
@@ -0,0 +1,78 @@
+#[warn(non_snake_case)]
+
+#[derive(Debug)]
+struct Natural {
+    n : i32
+}
+impl Natural {
+    fn new() -> Natural {
+        return Natural { n : 2 };
+    }
+
+    fn next(&mut self) -> i32 {
+        let v = self.n;
+        self.n = self.n + 1;
+        return v;
+    }
+}
+
+#[derive(Debug)]
+struct Filter {
+    n : i32,
+    next : Option<Box<Filter>>,
+}
+
+impl Filter {
+    fn new(n : i32) -> Filter {
+        Filter {
+            n : n,
+            next : None,
+        }
+    }
+
+    fn acceptAndAdd(&mut self, n : i32) -> bool {
+        let mut filter: &Filter = self;
+        let upto : i32 = (n as f64).sqrt() as i32;
+        loop {
+                if (n % filter.n) == 0 {
+                    return false;
+                }
+                if filter.n > upto {
+                    break;
+                }
+                match &filter.next {
+                    None => break,
+                    Some(go) => filter = &go,
+                }
+        }
+        let f = Filter::new(n);
+        Filter::append(self, f);
+        return true;
+    }
+
+    fn append(at : &mut Filter, f : Filter) {
+        loop {
+            match &at.next {
+                None => {
+                    at.next = Some(Box::new(f));
+                    break;
+                },
+                Some(n) => {
+                    at = n;
```

Comment:
The compiler gives following error here:
```
sieve/rust$ rustc sieve.rs 
error[E0308]: mismatched types
  --> sieve.rs:61:26
   |
61 |                     at = n;
   |                          ^ types differ in mutability
   |
   = note: expected mutable reference `&mut Filter`
                      found reference `&Box<Filter>`
```
what's the right way to implement `append` to the end of linked list in Rust?